File: Managed Source/Arsenal.ImageMounter/Reflection/SequenceComparer.vb
Imports System.Runtime.InteropServices

Namespace Reflection

    <ComVisible(False)>
    Public NotInheritable Class SequenceEqualityComparer(Of T)
        Implements IEqualityComparer(Of IEnumerable(Of T))

        Public Property ItemComparer As IEqualityComparer(Of T)

        Public Sub New(comparer As IEqualityComparer(Of T))
            _ItemComparer = comparer
        End Sub

        Public Sub New()
            _ItemComparer = EqualityComparer(Of T).Default
        End Sub

        Public Overloads Function Equals(x As IEnumerable(Of T), y As IEnumerable(Of T)) As Boolean Implements IEqualityComparer(Of IEnumerable(Of T)).Equals
            Return x.SequenceEqual(y, _ItemComparer)
        End Function

        Public Overloads Function GetHashCode(obj As IEnumerable(Of T)) As Integer Implements IEqualityComparer(Of IEnumerable(Of T)).GetHashCode
            Dim result As Integer
            For Each item In obj
                result = result Xor _ItemComparer.GetHashCode(item)
            Next
            Return result
        End Function

    End Class

    <ComVisible(False)>
    Public NotInheritable Class SequenceComparer(Of T)
        Implements IComparer(Of IEnumerable(Of T))

        Public Property ItemComparer As IComparer(Of T)

        Public Sub New(comparer As IComparer(Of T))
            _ItemComparer = comparer
        End Sub

        Public Sub New()
            _ItemComparer = Comparer(Of T).Default
        End Sub

        Public Function Compare(x As IEnumerable(Of T), y As IEnumerable(Of T)) As Integer Implements IComparer(Of IEnumerable(Of T)).Compare
            Dim value As Integer
            Using enumx = x.GetEnumerator()
                Using enumy = y.GetEnumerator()
                    While enumx.MoveNext() AndAlso enumy.MoveNext()
                        value = _ItemComparer.Compare(enumx.Current, enumy.Current)
                        If value <> 0 Then
                            Exit While
                        End If
                    End While
                End Using
            End Using
            Return value
        End Function
    End Class

End Namespace

File: Managed Source/Arsenal.ImageMounter/Reflection/SequenceComparer.cs
//  
//  Copyright (c) 2012-2025, Arsenal Consulting, Inc. (d/b/a Arsenal Recon) <http://www.ArsenalRecon.com>
//  This source code and API are available under the terms of the Affero General Public
//  License v3.
// 
//  Please see LICENSE.txt for full license terms, including the availability of
//  proprietary exceptions.
//  Questions, comments, or requests for clarification: http://ArsenalRecon.com/contact/
// 

using System;
using System.Collections.Generic;
using System.Linq;
using System.Runtime.InteropServices;



namespace Arsenal.ImageMounter.Reflection;

[ComVisible(false)]
public sealed class SequenceEqualityComparer<T> : IEqualityComparer<IEnumerable<T>>
{

    public IEqualityComparer<T> ItemComparer { get; set; }

    public SequenceEqualityComparer(IEqualityComparer<T> comparer)
    {
        ItemComparer = comparer;
    }

    public SequenceEqualityComparer()
    {
        ItemComparer = EqualityComparer<T>.Default;
    }

    public bool Equals(IEnumerable<T>? x, IEnumerable<T>? y)
        => ReferenceEquals(x, y)
        || (x is not null && y is not null && x.SequenceEqual(y, ItemComparer));

    public int GetHashCode(IEnumerable<T> obj)
    {
        var result = new HashCode();
        foreach (var item in obj)
        {
            result.Add(item, ItemComparer);
        }

        return result.ToHashCode();
    }
}

[ComVisible(false)]
public sealed class SequenceComparer<T> : IComparer<IEnumerable<T>>
{

    public IComparer<T> ItemComparer { get; set; }

    public SequenceComparer(IComparer<T> comparer)
    {
        ItemComparer = comparer;
    }

    public SequenceComparer()
    {
        ItemComparer = Comparer<T>.Default;
    }

    public int Compare(IEnumerable<T>? x, IEnumerable<T>? y)
    {
        if (ReferenceEquals(x, y))
        {
            return 0;
        }

        if (x is null)
        {
            return 1;
        }

        if (y is null)
        {
            return -1;
        }

        var value = 0;
        using var enumx = x.GetEnumerator();
        using var enumy = y.GetEnumerator();
        while (enumx.MoveNext() && enumy.MoveNext())
        {
            value = ItemComparer.Compare(enumx.Current, enumy.Current);
            if (value != 0)
            {
                break;
            }
        }

        return value;
    }
}